### PR TITLE
[2367] validate all confirm pages

### PR DESCRIPTION
**Pull request by stellenberger**
### Context

- [This trello ticket](https://trello.com/c/hB53BUOj/2367-l-validate-all-confirm-pages)

### Changes proposed in this pull request

- Moved the review summary outside `ApplyApplication` namespace
- Changed the `MissingDataView` to have the same API as `ApplyInvalidDataView`, some renaming of methods that run the same logic
- Changed the `ConfirmationController` to validate the forms if the user tries to click "I have completed this section"
- If the form is not valid, it re-renders the show view and populates the `MissingDataView` with an error message (red instead of blue)

### Guidance to review

- Make a draft trainee
- Go to `https://your_domain_here/trainees/{trainee.slug}/personal-details/confirm` directly from the URL bar
- You will be met with a bunch of blue errors with missing data
- Now, click `"I have completed this section"`
- The blue error box will be replaced with the red error messages.

### CODE REVIEW

File: app/forms/personal_details_form.rb [hunk 1]
```diff
@@ -41,7 +41,7 @@ def date_of_birth
   end
 
   def save!
-    if valid?
+    if valid?(:save)
```
Comment (jhenderson) [inline app/forms/personal_details_form.rb:44]: Not sure about adding the `:save` context. Does it change the behaviour? The specs all pass if I remove it. The context argument is usually `:create`, `:update` or a custom context.

File: app/components/review_summary/view.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    def invalid_data_exists?
+      form_is_valid = form.valid?
+
+      form.errors.clear unless errors_captured?
+
+      !form_is_valid && invalid_data_view.invalid_data?
+    end
+
+    def errors_captured?
+      @has_errors
```
Comment (jhenderson) [inline app/components/review_summary/view.rb:33]:
```suggestion
     form&.errors&.any?
```

File: app/components/review_summary/view.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module ReviewSummary
+  class View < GovukComponent::Base
+    renders_one :header
+
+    def initialize(form:, trainee:, data_view: ApplyInvalidDataView.new(trainee.apply_application))
+      @has_errors = form.respond_to?(:errors) ? form.errors.any? : false
```
Comment (jhenderson) [inline app/components/review_summary/view.rb:8]: You can get rid of this line and use what's suggested below for `errors_captured?`

File: app/components/review_summary/view.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    end
+
+    def summary_component
+      @summary_component ||= errors_captured? ? ErrorSummary::View : InformationSummary::View
+    end
+
+  private
+
+    def invalid_data_exists?
+      form_is_valid = form.valid?
```
Comment (jhenderson) [inline app/components/review_summary/view.rb:25]:
```suggestion
      form.errors.clear unless errors_captured?

      !form.valid? && invalid_data_view.invalid_data?
```

File: app/controllers/trainees/confirm_details_controller.rb [hunk 5]
```diff
@@ -12,24 +12,31 @@ def show
...
     def update
-      form_klass.new(trainee).save! unless trainee.draft?
-
-      toggle_trainee_progress_field if trainee.draft?
-
-      flash[:success] = "Trainee #{flash_message_title} updated"
-
-      redirect_to page_tracker.last_origin_page_path || trainee_path(trainee)
+      if form.valid?
+        form_klass.new(trainee).save! unless trainee.draft?
```
Comment (jhenderson) [inline app/controllers/trainees/confirm_details_controller.rb:23]: You've created a new method called `form`. You can take advantage of this:

```suggestion
        form.save! unless trainee.draft?
```

File: app/controllers/trainees/confirm_details_controller.rb [hunk 6]
```diff
@@ -12,24 +12,31 @@ def show
...
+        toggle_trainee_progress_field if trainee.draft?
+        flash[:success] = "Trainee #{flash_message_title} updated"
+        redirect_to page_tracker.last_origin_page_path || trainee_path(trainee)
+      else
+        @confirmation_component = component_klass.new(data_model: trainee.draft? ? trainee : form_klass.new(trainee), has_errors: true)
```
Comment (jhenderson) [inline app/controllers/trainees/confirm_details_controller.rb:28]:
```suggestion
        @confirmation_component = component_klass.new(data_model: trainee.draft? ? trainee : form, has_errors: true)
```

File: app/components/review_summary/view.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      @trainee = trainee
+      @invalid_data_view = data_view
+    end
+
+    def render?
+      errors_captured? || invalid_data_exists?
+    end
+
+    def summary_component
+      @summary_component ||= errors_captured? ? ErrorSummary::View : InformationSummary::View
```
Comment (jhenderson) [inline app/components/review_summary/view.rb:19]: You could put this in the `initialize` method and create a public accessor method. Saves you from having to create a separate method and then having to memoise the result.

**Review by jhenderson — CHANGES_REQUESTED** [review]
Works as expected - good job. Left some code feedback.

File: app/forms/personal_details_form.rb (hunk 1 above)
Comment (darokel) [reply app/forms/personal_details_form.rb:44]: this is a remnant, it can go in the bin

**Comment by jhenderson** [comment]
[[L] Validate all confirm pages](https://trello.com/c/hB53BUOj/2367-l-validate-all-confirm-pages)

**Comment by bbelward** [comment]
Working for `personal-details` but not for diversity information (ignore the second part of the gif, was just trying to get to the degree confirmation page) 
2367diversityinfo

File: app/controllers/trainees/confirm_details_controller.rb [hunk 8]
```diff
@@ -52,12 +61,8 @@ def form_klass
       end
     end
 
-    def form_instance
-      trainee_section_key == SCHOOLS_KEY ? form_klass.new(trainee, non_search_validation: true) : form_klass.new(trainee)
-    end
-
     # Returns the route that the confirm path is nested under for each confirm path
-    # eg /trainees/HxA35kmiNdtwNGnWYjr6FbJZ/funding/confirm returns 'funding'
+    # eg /trainees/<slug>/funding/confirm returns 'funding'
```
Comment (archferns) [inline app/controllers/trainees/confirm_details_controller.rb:65]: :+1:

**Review by archferns — APPROVED** [review]

**Comment by bbelward** [comment]
So, checked all sections and is working as expected, however the error messages and the visual design is slightly different for degrees. Might be worth carding up separately, @edwardhorsford are you aware of this? 
2367degreedetails

**Review by jhenderson — APPROVED** [review]

**Comment by edwardhorsford** [comment]
There were a few bugs with this - [I've carded up here](https://trello.com/c/HVRzOZEw/2517-snags-with-validated-confirm-pages).

I agree degree is funky. Not entirely sure what to do about it - I guess at the least we can raise a design card.

**Pull request merged**